src/main/scala/akka/persistence/cassandra/query/scaladsl/CassandraReadJournal.scala

```diff
@@ -0,0 +1,223 @@
+package akka.persistence.cassandra.query.scaladsl
+
+import java.net.URLEncoder
+import java.time.LocalDate
+import java.time.ZoneOffset
+import java.util.UUID
+import akka.actor.ExtendedActorSystem
+import akka.persistence.cassandra.journal.CassandraJournal
+import akka.persistence.cassandra.journal.CassandraJournalConfig
+import akka.persistence.cassandra.journal.CassandraStatements
+import akka.persistence.cassandra.query.CassandraReadJournalConfig
+import akka.persistence.cassandra.query.CassandraReadStatements
+import akka.persistence.cassandra.query.EventsByTagPublisher
+import akka.persistence.cassandra.query.UUIDEventEnvelope
+import akka.persistence.query._
+import akka.persistence.query.scaladsl._
+import akka.stream.ActorAttributes
+import akka.stream.scaladsl.Source
+import akka.util.ByteString
+import com.datastax.driver.core.PreparedStatement
+import com.datastax.driver.core.utils.UUIDs
+import com.typesafe.config.Config
+import akka.persistence.cassandra.retry
+import com.datastax.driver.core.Session
+
+object CassandraReadJournal {
+  /**
+   * The default identifier for [[CassandraReadJournal]] to be used with
+   * `akka.persistence.query.PersistenceQuery#readJournalFor`.
+   *
+   * The value is `"cassandra-query-journal"` and corresponds
+   * to the absolute path to the read journal configuration entry.
+   */
+  final val Identifier = "cassandra-query-journal"
+}
+
+/**
+ * Scala API `akka.persistence.query.scaladsl.ReadJournal` implementation for Cassandra.
+ *
+ * It is retrieved with:
+ * {{{
+ * val queries = PersistenceQuery(system).readJournalFor[CassandraReadJournal](CassandraReadJournal.Identifier)
+ * }}}
+ *
+ * Corresponding Java API is in [[akka.persistence.cassandra.query.javadsl.CassandraReadJournal]].
+ *
+ * Configuration settings can be defined in the configuration section with the
+ * absolute path corresponding to the identifier, which is `"cassandra-query-journal"`
+ * for the default [[CassandraReadJournal#Identifier]]. See `reference.conf`.
+ */
+class CassandraReadJournal(system: ExtendedActorSystem, config: Config)
+  extends ReadJournal
+  with EventsByTagQuery
+  with CurrentEventsByTagQuery {
+
+  private val writePluginConfig = new CassandraJournalConfig(system.settings.config.getConfig(config.getString("write-plugin")))
+  private val queryPluginConfig = new CassandraReadJournalConfig(config, writePluginConfig)
+
+  private class CassandraSession {
+
+    val session: Session = connect()
+
+    private def connect(): Session = {
+      val cluster = writePluginConfig.clusterBuilder.build
+      retry(writePluginConfig.connectionRetries + 1, writePluginConfig.connectionRetryDelay.toMillis) {
```

Comment:
yes, I didn't want to change too much, since I know it was in progress elsewhere.